drivers/interrupt_controller/intc_xlnx.c

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (c) 2023 - 2024 Advanced Micro Devices, Inc. (AMD)
+ * Copyright (c) 2023 Alp Sayin <alpsayin@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * This file implements AXI Interrupt Controller (INTC)
+ * For more details about the INTC see PG 099
+ *
+ * The functionality has been based on intc_v3_12 package
+ *
+ * Right now the implementation:
+ *  - does not support fast interrupt mode
+ *  - does not support Cascade mode
+ *  - does not support XIN_SVC_SGL_ISR_OPTION
+ */
+
+#include <errno.h>
+#include <zephyr/arch/cpu.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/interrupt_controller/intc_xlnx.h>
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/sys_clock.h>
+
+LOG_MODULE_REGISTER(xlnx_intc);
+
+#define DT_DRV_COMPAT xlnx_xps_intc_1_00_a
+
+#define BASE_ADDRESS	 DT_INST_REG_ADDR(0)
+#define INTC_REG(offset) (uint32_t *)(BASE_ADDRESS + offset)
+
+#define xlnx_intc_read(offset)	      sys_read32(BASE_ADDRESS + offset)
+#define xlnx_intc_write(data, offset) sys_write32(data, BASE_ADDRESS + offset)
+
+#define XIN_SVC_SGL_ISR_OPTION	1UL
+#define XIN_SVC_ALL_ISRS_OPTION 2UL
+
+#define XIN_ISR_OFFSET 0x0  /* Interrupt Status Register */
+#define XIN_IPR_OFFSET 0x4  /* Interrupt Pending Register */
+#define XIN_IER_OFFSET 0x8  /* Interrupt Enable Register */
+#define XIN_IAR_OFFSET 0xc /* Interrupt Acknowledge Register */
+#define XIN_SIE_OFFSET 0x10 /* Set Interrupt Enable Register */
+#define XIN_CIE_OFFSET 0x14 /* Clear Interrupt Enable Register */
+#define XIN_IVR_OFFSET 0x18 /* Interrupt Vector Register */
+#define XIN_MER_OFFSET 0x1C /* Master Enable Register */
+#define XIN_IMR_OFFSET 0x20 /* Interrupt Mode Register, only for Fast Interrupt */
+#define XIN_IVAR_OFFSET	0x100 /* Interrupt Vector Address Register, only for Fast Interrupt */
+
+/* Bit definitions for the bits of the MER register */
+#define XIN_INT_MASTER_ENABLE_MASK	BIT(0)
+#define XIN_INT_HARDWARE_ENABLE_MASK	BIT(1) /* once set cannot be cleared */
+
+struct xlnx_intc_state {
+	bool is_ready;	 /* Device is initialized and ready */
+	bool is_started; /* Device has been started */
+};
+
+static struct xlnx_intc_state intc_state = {
+	.is_ready = false,
+	.is_started = false,
+};
```

Comment:
This doesn't support multiple instances?

Reply:
Definitely you can wire multiple instances but over years I didn't see Microblaze design with Linux where users use more then 32 irq lines. But definitely you can connect more than one instance of it.